update openstack cloud schema to include os_cacert

hmlanigan opened the pull request:
## Description of change

Update add-cloud to allow default values of the cloud schema's to get
values from environment variables.  Add environment variables and a new
get cloud ca certificate to the openstack cloud schema.  Add ca-certificates
to cloud schema use for validation.

## QA steps

Run juju add-cloud for openstack.  Should query for an optional cloud ca cert.
1. try without a cert
2. try with a good cert file
3. try with a bad cert file - should create the cloud but Ignore the cert.

Source a novarc file, including a OS_CACERT definition, run juju add-cloud for openstack.
1. notice that the auth url is now a default option.
2. notice that the os cacert is now a default option.

Check that the new cloud definition in ~/.local/share/juju/clouds.yaml contains  ca-certificates: ...

Try juju add-cloud -f with a yaml definition including the ca-certificates, no schema validation error should be printed on the name.

## Documentation changes

Not yet, a discourse update will happen with all of the OS_CACERT changes are made.

## Bug reference

https://bugs.launchpad.net/juju/+bug/1777897
Partially addresses https://bugs.launchpad.net/juju/+bug/1722580

hmlanigan commented:
![](https://github.trello.services/images/mini-trello-icon.png) [cloud ca-cert via interactive juju add-cloud](https://trello.com/c/XBcblQaG/414-cloud-ca-cert-via-interactive-juju-add-cloud)

hmlanigan commented:
!!build!!

mitechie commented:
Interactive didn't work for me. It output the optional cacert but went stright into the Auth types?

https://pastebin.canonical.com/p/kBwRQv3bDk/

I never hit enter or anything to deal wit hthe filename and I'd expect the authtypes to have a newline after it to line up with other questions.

mitechie commented on cmd/juju/cloud/add_test.go:
```diff
@@ -757,3 +706,260 @@ func (s *addSuite) TestInvalidCredentialMessage(c *gc.C) {
+		/* Select cloud type: */ "openstack\n" +
+		/* Enter a name for your openstack cloud: */ "os1\n" +
+		/* Enter the API endpoint url for the cloud []: */ "http://myopenstack\n" +
+		/* Enter the filename of the CA certificate to access OpenStack cloud (optional) [] */ "\n" +
```
I know these are the tests but wording suggestion

    If your cloud requires a CA certificate please provide the path to the certificate (optional) [none]:

hmlanigan replied:
There isn't a good way to get rid of a space at the beginning:
"Enter , if required, the path to a CA"... 

Sorry I wasn't more specific, but "Enter " gets tacked to the beginning on the string I provide..

Does "Enter if required, the path to a CA" work for you?

mitechie replied:
no, I don't think that's proper punctuation there. #$@$R#@ this code is here just to irritate me I'm beginning to think.

mitechie replied:
Enter a path to the CA certificate for your cloud if one is required to access it. (optional) [none]

I hate that I can't think of a better way to get the optional-ity of the question towards the front of the line given the stupid constraints of this code.

hmlanigan replied:
updated.

mitechie commented:
I also note when interactive adding an openstack cloud you get asked for the "API Endpoint url" for the cloud but then you get asked about the region and the region url. Are these different urls? 

In checking our wiki on internal openstack the config is a single url for the region?

mitechie commented:
Interesting, when I ran juju add-cloud without giving it a name I got the above pastebin. When I ran 

    juju add-cloud testingstack3

it did pause and ask me about the CA cert properly.

mitechie commented:
ok, I do see the region url defaults to "[use cloud api url]" can this actually be the value vs the generic text?

mitechie commented:
We should fail if the cert won't load right away as if we typed in a wrong string for the auth type or a ping that failed

https://pastebin.canonical.com/p/rwZQDcM5pX/

mitechie commented:
I guess I can see the delayed response on the invalid cert as we're not reading it yet...if we can read it right away that'd be great. If not, shoudn't it be an error vs a warning and we should not add the cloud anyway?

https://pastebin.canonical.com/p/rs3P4gnSHx/

mitechie commented:
Awesome when you have the sourced file how the API url is prefilled with the real value like that. That's great.

hmlanigan commented:
Side effect of the environment variables is that they aren't verified if the FormatURI as the value would be without it.   Risk should be minimal, as with the OpenStack case, the values have been verified by use when interacting with their OpenStack cloud.

hmlanigan commented:
Changed the default of Y/n to N/y when asking "Enter another... "
Changed to validate the certificate file when filename provided.

Comment on "[use cloud api url]" outside of scope for this PR.  Please file a bug.
We also need a bug to strip extra newlines and spaces from the input during add-cloud interactive.  Specifying "openstack   " for cloud type fails too.

hmlanigan commented on cmd/juju/cloud/add_test.go:
```diff
@@ -65,6 +65,10 @@ func (f *fakeCloudMetadataStore) WritePersonalCloudMetadata(cloudsMap map[string
 
 func (f *fakeCloudMetadataStore) ParseOneCloud(data []byte) (cloudfile.Cloud, error) {
 	results := f.MethodCall(f, "ParseOneCloud", data)
+	if len(results) != 2 {
```
This was annoying me during testing, if the data doesn't match, the test panics.  Put in for debugging, since the printed []byte is not helpful to reading strings to find what's wrong.

SimonRichardson commented on cmd/juju/cloud/add.go:
```diff
@@ -235,6 +264,45 @@ func (c *AddCloudCommand) runInteractive(ctxt *cmd.Context) error {
+		return "", nil, err
+	}
+	var certificate string
+	filename, ok := vals[cloud.CertFilenameKey].(string)
```
This will panic if the `cloud.CertFilenameKey` is not there, because you're using the type assertion as well. If we're 100% sure that this will be fine, ignore my comment.

hmlanigan replied:
Updating:  I can't guarantee the key in the map, but I can guarantee if the key in the map, the value is a string.

hmlanigan commented:
$$merge$$
